homeassistant/components/dremel_3d_printer/services.py

```diff
@@ -0,0 +1,119 @@
+        _LOGGER.warning("Not a file: %s", filepath)
+        return False
+    return True
+
+
+async def async_setup_services(hass: HomeAssistant) -> None:
+    """Set up services for the Dremel 3D Printer integration."""
+
+    def get_api(service: ServiceCall) -> Dremel3DPrinter:
+        """Return the host ip of a Dremel 3D Printer device."""
+        device_id = service.data[ATTR_DEVICE_ID]
+        dev_reg = device_registry.async_get(hass)
+        if (device_entry := dev_reg.async_get(device_id)) is None:
+            raise vol.Invalid("Invalid device ID")
+        config_list = list(device_entry.config_entries)
+        if len(config_list) == 0:
+            raise vol.Invalid("No config entries for device ID")
+        config_entry = list(device_entry.config_entries)[0]
+        return hass.data[DOMAIN][config_entry].api
+
+    async def print_job(service: ServiceCall) -> None:
+        """Service to start a printing job."""
+        api = get_api(service)
+        filepath = service.data.get(ATTR_FILEPATH)
+        url = service.data.get(ATTR_URL)
+        try:
+            if filepath is not None:
+                if not file_exists(hass, filepath):
+                    raise RuntimeError("File does not exist.")
+                if not filepath.lower().endswith(".gcode"):
+                    raise RuntimeError("File is not a gcode file.")
+                result = await hass.async_add_executor_job(
+                    api.start_print_from_file, filepath
+                )
+
+            if url is not None:
+                result = await hass.async_add_executor_job(
+                    api.start_print_from_url, url
+                )
+
+            hass.bus.async_fire(
+                EVENT_DATA_NEW_PRINT_STATS,
+                result,
+            )
+        except Exception as exc:  # pylint: disable=broad-except
+            _LOGGER.error(str(exc))
+
+    async def pause_job(service: ServiceCall) -> None:
+        """Service to pause a printing job."""
+        api = get_api(service)
+        await hass.async_add_executor_job(api.pause_print)
+
+    async def resume_job(service: ServiceCall) -> None:
+        """Service to resume a printing job."""
+        api = get_api(service)
+        await hass.async_add_executor_job(api.resume_print)
+
+    async def stop_job(service: ServiceCall) -> None:
+        """Service to stop a printing job."""
+        api = get_api(service)
+        await hass.async_add_executor_job(api.stop_print)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_PRINT_JOB, print_job, schema=SERVICE_PRINT_JOB_SCHEMA
+    )
+    hass.services.async_register(
+        DOMAIN, SERVICE_PAUSE_JOB, pause_job, schema=SERVICE_COMMON_JOB_SCHEMA
+    )
+    hass.services.async_register(
+        DOMAIN, SERVICE_RESUME_JOB, resume_job, schema=SERVICE_COMMON_JOB_SCHEMA
+    )
+    hass.services.async_register(
+        DOMAIN, SERVICE_STOP_JOB, stop_job, schema=SERVICE_COMMON_JOB_SCHEMA
+    )
```

Comment:
These all seem to work with the button entity? If so, this can wait for a future PR.